bee-network/src/peer/list.rs

```diff
@@ -0,0 +1,469 @@
+                        relation: PeerRelation::Known,
+                    },
+                    PeerState::new_disconnected(),
+                ),
+            )
+        }));
+
+        Self {
+            local_id,
+            local_addrs: HashSet::with_capacity(LOCAL_ADDRS_INITIAL_CAP),
+            peers: p,
+            banned_peers: HashSet::default(),
+            banned_addrs: HashSet::default(),
+        }
+    }
+
+    pub fn insert_peer(&mut self, peer_id: PeerId, peer_info: PeerInfo) -> Result<(), (PeerId, PeerInfo, Error)> {
+        if self.contains(&peer_id) {
+            return Err((peer_id, peer_info, Error::PeerIsAdded(peer_id)));
+        }
+
+        // Since we already checked that such a `peer_id` is not yet present, the returned value is always `None`.
+        let _ = self.peers.insert(peer_id, (peer_info, PeerState::new_disconnected()));
+
+        Ok(())
+    }
+
+    pub fn insert_local_addr(&mut self, addr: Multiaddr) -> Result<(), (Multiaddr, Error)> {
+        if self.local_addrs.contains(&addr) {
+            return Err((addr.clone(), Error::AddressIsAdded(addr)));
+        }
+
+        let _ = self.local_addrs.insert(addr);
+
+        Ok(())
+    }
+
+    pub fn remove(&mut self, peer_id: &PeerId) -> Result<PeerInfo, Error> {
+        let (info, _) = self
+            .peers
+            .remove(peer_id)
+            .ok_or_else(|| Error::PeerNotPresent(*peer_id))?;
+
+        Ok(info)
+    }
+
+    pub fn contains(&self, peer_id: &PeerId) -> bool {
+        self.peers.contains_key(peer_id)
+    }
+
+    pub fn info(&self, peer_id: &PeerId) -> Result<PeerInfo, Error> {
+        self.peers
+            .get(peer_id)
+            .ok_or_else(|| Error::PeerNotPresent(*peer_id))
+            .map(|(peer_info, _)| peer_info.clone())
+    }
+
+    pub fn len(&self) -> usize {
+        self.peers.len()
+    }
+
+    pub fn update_info<U>(&mut self, peer_id: &PeerId, mut update: U) -> Result<(), Error>
+    where
+        U: FnMut(&mut PeerInfo),
+    {
+        let (info, _) = self
+            .peers
+            .get_mut(peer_id)
+            .ok_or_else(|| Error::PeerNotPresent(*peer_id))?;
+
+        update(info);
+
+        Ok(())
+    }
+
+    pub fn update_state<U>(&mut self, peer_id: &PeerId, mut update: U) -> Result<Option<GossipSender>, Error>
+    where
+        U: FnMut(&mut PeerState) -> Option<GossipSender>,
+    {
+        let (_, state) = self
+            .peers
+            .get_mut(peer_id)
+            .ok_or_else(|| Error::PeerNotPresent(*peer_id))?;
+
+        Ok(update(state))
+    }
+
+    pub fn is<P>(&self, peer_id: &PeerId, predicate: P) -> Result<bool, Error>
+    where
+        P: Fn(&PeerInfo, &PeerState) -> bool,
+    {
+        self.peers
+            .get(peer_id)
+            .ok_or_else(|| Error::PeerNotPresent(*peer_id))
+            .map(|(info, state)| predicate(info, state))
+    }
+
+    pub fn iter_if<P>(&self, predicate: P) -> impl Iterator<Item = (PeerId, String)>
```

Comment:
maybe this should be called `filter` or something similar

Reply:
yup, that would be more in line with `std`. Agree!